nat-lab/tests/utils/network_switcher/network_switcher_windows.py

```diff
@@ -1,16 +1,68 @@
 import config
+import re
 from .network_switcher import NetworkSwitcher
+from dataclasses import dataclass
+from typing import List, Optional
 from utils.connection import Connection
 from utils.process import ProcessExecError
 
-VAGRANT_DEFAULT_IF = "Ethernet"  # 10.0.2.0/24
-PRIMARY_IF = "Ethernet 2"  # 10.55.0.11
-SECONDARY_IF = "Ethernet 3"  # 10.66.0.11
+
+@dataclass
+class Interface:
+    def __init__(self, name: str, ipv4: str) -> None:
+        self.name = name
+        self.ipv4 = ipv4
+
+
+class ConfiguredInterfaces:
+    def __init__(self, default: str, primary: str, secondary: str) -> None:
+        self.default = default
+        self.primary = primary
+        self.secondary = secondary
+
+    @staticmethod
+    async def find(connection: Connection) -> "ConfiguredInterfaces":
+        interfaces = await ConfiguredInterfaces.get_network_interfaces(connection)
+
+        def find_interface(prefix: str) -> str:
+            for interface in interfaces:
+                if interface.ipv4.startswith(prefix):
+                    return interface.name
+            assert False, f"interface not found with prefix `{prefix}`, {interfaces}"
+
+        return ConfiguredInterfaces(
+            find_interface(config.LIBVIRT_MANAGEMENT_NETWORK_PREFIX),
+            find_interface(config.PRIMARY_VM_NETWORK_PREFIX),
+            find_interface(config.SECONDARY_VM_NETWORK_PREFIX),
+        )
+
+    @staticmethod
+    async def get_network_interfaces(connection: Connection) -> List[Interface]:
+        process = await connection.create_process(
+            ["netsh", "interface", "ipv4", "show", "addresses"]
+        ).execute()
+
+        matches = re.findall(
+            r"Configuration for interface \"(.*)\"[\s\S]*?IP Address:\s*([\d.]*)",
+            process.get_stdout(),
+        )
+
+        result: List[Interface] = []
+        for match in matches:
+            result.append(Interface(match[0], match[1]))
+
+        return result
 
 
 class NetworkSwitcherWindows(NetworkSwitcher):
     def __init__(self, connection: Connection) -> None:
         self._connection = connection
+        self._interfaces: Optional[ConfiguredInterfaces] = None
```

Comment:
Are these interfaces supposed to change during the test run? Maybe you could initialize them in the constructor and then just use?

Reply:
Nope, they aren't going to change during test run. I wish I could initialize them, but constructors can't be async :( Doing this lazy initialization is one of solutions, and the other solution would be to add "post init" method. Now that I think about it, adding post init method should be possible without changing the abstract network switcher interface.

Reply:
But I'm not sure what would be gained from that, because `_interfaces` would still be optional, and would need to be asserted before each use to satisfy the type checker. So the end result would be the same :(

Reply:
Well, that's a pain. Maybe it could be initialized by with a list of Interfaces?

Reply:
I think that would work, I will work something out and you can take look at it again.